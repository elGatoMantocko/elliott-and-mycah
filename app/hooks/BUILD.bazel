
load("@npm//@bazel/typescript:index.bzl", "ts_library")
load("@npm//eslint:index.bzl", "eslint")

filegroup(
    name = "source_files",
    srcs = glob(["*.tsx"]),
)

eslint(
    name = "eslint",
    data = [
        "//:.eslintrc",
        ":source_files",
        "@npm//:node_modules",
    ],
    args = ["**/*.tsx"],
)

ts_library(
    name = "ts_library",
    srcs = [":source_files"],
    deps = [
        "@npm//@material-ui/core",
        "@npm//@types",
        "@npm//react",
    ],
    visibility = ["//visibility:public"],
)