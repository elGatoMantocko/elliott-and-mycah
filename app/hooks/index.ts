export * from './result';
export * from './scrollToTop';
export * from './serviceWorker';
export * from './theme';
