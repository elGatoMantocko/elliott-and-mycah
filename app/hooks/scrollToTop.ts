import { useEffect } from 'react';
import { useLocation } from 'react-router-dom';

/**
 * Hook that will automatically make a browser scroll to the `0` position when
 * the path changes inside of a router.
 */
export const useScrollToTop = () => {
  const { pathname } = useLocation();

  useEffect(() => {
    window.scrollTo(0, 0);
  }, [pathname]);
};
