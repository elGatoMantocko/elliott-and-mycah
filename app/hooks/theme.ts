import { createTheme, PaletteOptions, useMediaQuery } from '@mui/material';
import { useCallback, useMemo } from 'react';

export const useCustomTheme = () => {
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)');

  const mode = useMemo(() => (prefersDarkMode ? 'dark' : 'light'), [prefersDarkMode]);

  const spacing = useCallback(
    (factor: number | string): number | string =>
      typeof factor === 'number' ? `${0.5 * factor}rem` : factor,
    [],
  );

  const palette = useMemo(
    (): PaletteOptions | undefined => ({
      mode,
      // TODO: play with dark mode but this site is designed for light mode
      primary: {
        dark: 'rgba(175, 126, 203, 1)',
        main: 'rgba(139, 77, 174, 1)',
        light: 'rgba(114, 78, 135, 1)',
        contrastText: '#fff',
      },
      secondary: {
        dark: 'rgba(104, 177, 104, 1)',
        main: 'rgba(31, 134, 31, 1)',
        light: 'rgba(45, 92, 45, 1)',
        contrastText: '#fff',
      },
      error: {
        dark: '#e57373',
        main: '#f44336',
        light: '#d32f2f',
        contrastText: '#000',
      },
      action: {
        // default disabled has a lot of opacity and that's stupid
        disabled: 'rgba(159, 159, 159, 1)',
        disabledBackground: 'rgba(226, 226, 226, 1)',
      },
    }),
    [mode],
  );

  return createTheme({ spacing, palette });
};
