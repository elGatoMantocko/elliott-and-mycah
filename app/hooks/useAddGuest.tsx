import { useCallableResult } from './useResult';

export const useAddGuest = () =>
  useCallableResult(
    (
      firstName: string,
      lastName: string,
      yesNo?: boolean,
      foodChoice?: 'chicken' | 'fish' | 'veg',
    ) =>
      fetch('https://sheet.best/api/sheets/4976d0ec-48c2-40ef-b4d3-740560714fb1', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ firstName, lastName, foodChoice, yesNo }),
      }),
  );
