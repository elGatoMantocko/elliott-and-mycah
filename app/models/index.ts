export * from './guest';
export * from './state';
