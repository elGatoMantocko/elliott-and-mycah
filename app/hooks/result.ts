import { DependencyList, useCallback, useEffect, useState } from 'react';

import { State, state } from '../models';

export enum ResultState {
  NotStarted = 'not-started',
  Pending = 'pending',
  Value = 'value',
  Error = 'error',
}

export type Result<T, U> =
  | State<ResultState.NotStarted>
  | State<ResultState.Pending>
  | State<ResultState.Value, T>
  | State<ResultState.Error, U>;

const PENDING: State<ResultState.Pending> = state(ResultState.Pending);
const NOT_STARTED: State<ResultState.NotStarted> = state(ResultState.NotStarted);

/**
 * Turns an unknown error into an {@link Error}
 * @param err unknown error
 * @returns a valid error object
 */
function handleUnknownError(err: unknown): Error {
  if (err instanceof Error) {
    return err;
  } else if (typeof err === 'string') {
    return new Error(err);
  } else if (typeof err === 'object') {
    return new Error(`encountered error in useCallableResult: ${JSON.stringify(err)}`);
  } else {
    return new Error('encountered unknown error in useCallableResult');
  }
}

/**
 * Function to get a result from an asynchronous function.
 * @param request asynchronous function to map to a result state
 * @param deps list of deps used in the request
 * @returns result state object
 */
export const useResult = <C extends () => Promise<unknown>, CValue extends Awaited<ReturnType<C>>>(
  request: C extends () => Promise<CValue> ? C : never,
  deps: DependencyList,
) => {
  const [result, setResult] = useState<Result<CValue, Error>>(NOT_STARTED);

  useEffect(() => {
    let isSubscribed = true;

    // only set state and make requests as long as this hook is still "mounted" to the dom
    // this is mostly just so we don't freak the tests out when things get unmounted
    if (isSubscribed) {
      setResult(PENDING);
      request()
        .then((data) => setResult(state(ResultState.Value, data)))
        .catch((err: unknown) => setResult(state(ResultState.Error, handleUnknownError(err))));
    }

    // if the `useResult` hook is torn down, we want to make sure we don't have any
    // lingering calls being made
    return () => {
      isSubscribed = false;
    };
    // The caller is defining the dependencies to this so we intentionally
    // want to ignore the exhaustive deps rule here.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...deps]);
  return result;
};

/**
 * Hook to get a callable function that sets the result state. Can be reset to `not-started`
 * with the 3rd option in the tuple.
 * @param request function to call
 * @returns tuple representing the `[result-state, callable, reset]` options
 */
export const useCallableResult = <CArgs extends unknown[], CResponse>(
  request: (...args: CArgs) => Promise<CResponse>,
): [Result<CResponse, Error>, (...args: CArgs) => void, () => void] => {
  const [result, setResult] = useState<Result<CResponse, Error>>(NOT_STARTED);

  const call = useCallback(
    async (...args: CArgs) => {
      setResult(PENDING);

      try {
        const t = await request(...args);

        setResult(state(ResultState.Value, t));
      } catch (err) {
        setResult(state(ResultState.Error, handleUnknownError(err)));
      }
    },
    [request],
  );

  return [result, call, () => setResult(NOT_STARTED)];
};
