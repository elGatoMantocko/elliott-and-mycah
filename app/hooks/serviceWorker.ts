import { useResult } from './result';

/**
 * Hook to unregister all service workers.
 * @returns true if any and all service workers were able to be unregistered
 */
export const useUnregisterServiceWorkers = () =>
  useResult(async () => {
    const registrations = await navigator.serviceWorker.getRegistrations();
    const results = await Promise.all(registrations.map((r) => r.unregister()));
    return results.length > 0 && results.every((didUnregister) => didUnregister);
  }, []);
