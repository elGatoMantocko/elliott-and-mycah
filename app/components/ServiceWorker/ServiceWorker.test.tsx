import { act, render } from '@testing-library/react';
import React from 'react';

import { ResultState } from '../../hooks';
import { State, state } from '../../models';
import { ServiceWorkerUnregistration } from '.';

jest.mock('../../hooks/serviceWorker', () => ({
  __esModule: true,
  useUnregisterServiceWorkers: jest.fn(),
}));

/**
 * Helper to mock a sw unregister result.
 * @param result that the unregister service worker hook returns
 */
async function mockUnregisterResult(
  result: State<ResultState, boolean | Error | undefined>,
): Promise<void> {
  const { useUnregisterServiceWorkers } = await import('../../hooks/serviceWorker');
  // See line 7 in this file. We are setting up the module mock for this
  // and we want to override the implementation in each test
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  useUnregisterServiceWorkers.mockImplementation(() => result);
}

it('should render a successful <Unregister />', async () => {
  await mockUnregisterResult(state(ResultState.Value, true));

  const el = render(
    <div data-testid="unregister-sw-success">
      <ServiceWorkerUnregistration />
    </div>,
  );

  expect(window.location.reload).toHaveBeenCalled();

  const unregistration = await el.findByTestId('unregister-sw-success');
  expect(unregistration).toMatchSnapshot('success-unregister');
});

it('should render a pending <Unregister />', async () => {
  await mockUnregisterResult(state(ResultState.Pending));

  const el = render(
    <div data-testid="unregister-sw-pending">
      <ServiceWorkerUnregistration />
    </div>,
  );

  expect(window.location.reload).toHaveBeenCalled();

  const unregistration = await el.findByTestId('unregister-sw-pending');
  expect(unregistration).toMatchSnapshot('pending-unregister');
});

it('should render an error <Unregister />', async () => {
  // this will trigger an error log in the tests, it's expected until I refactor the
  // logger to be injected into components
  await mockUnregisterResult(state(ResultState.Error, new Error('oops')));

  const el = render(
    <div data-testid="unregister-sw-err">
      <ServiceWorkerUnregistration />
    </div>,
  );

  expect(window.location.reload).toHaveBeenCalled();

  const unregistration = await el.findByTestId('unregister-sw-err');
  expect(unregistration).toMatchSnapshot('error-unregister');
});

it('should render a successful <Unregister /> with a snackbar', async () => {
  jest.useFakeTimers();

  await mockUnregisterResult(state(ResultState.Value, true));

  const ui = (
    <div data-testid="unregister-sw-success-hide-snack">
      <ServiceWorkerUnregistration />
    </div>
  );

  const el = render(ui);

  expect(window.location.reload).toHaveBeenCalled();

  expect(await el.findByTestId('unregister-sw-success-hide-snack')).toMatchSnapshot(
    'success-unregister',
  );

  await act(() => {
    jest.advanceTimersByTime(3000);
    el.rerender(ui);
  });

  expect(await el.findByTestId('unregister-sw-success-hide-snack')).toMatchSnapshot(
    'success-unregister-post-hide',
  );
});
