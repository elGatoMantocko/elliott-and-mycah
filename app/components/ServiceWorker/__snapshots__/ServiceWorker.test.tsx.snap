// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a pending <Unregister />: pending-unregister 1`] = `
<div
  data-testid="unregister-sw-pending"
/>
`;

exports[`should render a successful <Unregister /> with a snackbar: success-unregister 1`] = `
<div
  data-testid="unregister-sw-success-hide-snack"
>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter css-1t20nyk-MuiSnackbar-root"
    color="primary"
    role="presentation"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation6 MuiSnackbarContent-root css-1eqdgzv-MuiPaper-root-MuiSnackbarContent-root"
      direction="up"
      role="alert"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message css-1exqwzz-MuiSnackbarContent-message"
      >
        <div
          class="MuiBox-root css-1dhxv6h"
        >
          <div
            class="MuiBox-root css-oq1jtl"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ThumbUpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z"
              />
            </svg>
          </div>
          <div
            class="MuiBox-root css-1snmjsk"
          >
            Service worker loaded!
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render a successful <Unregister /> with a snackbar: success-unregister-post-hide 1`] = `
<div
  data-testid="unregister-sw-success-hide-snack"
>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter css-1t20nyk-MuiSnackbar-root"
    color="primary"
    role="presentation"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation6 MuiSnackbarContent-root css-1eqdgzv-MuiPaper-root-MuiSnackbarContent-root"
      direction="up"
      role="alert"
      style="opacity: 0; transform: scale(0.75, 0.5625); transition: opacity 195ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 130ms cubic-bezier(0.4, 0, 0.2, 1) 65ms;"
    >
      <div
        class="MuiSnackbarContent-message css-1exqwzz-MuiSnackbarContent-message"
      >
        <div
          class="MuiBox-root css-1dhxv6h"
        >
          <div
            class="MuiBox-root css-oq1jtl"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ThumbUpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z"
              />
            </svg>
          </div>
          <div
            class="MuiBox-root css-1snmjsk"
          >
            Service worker loaded!
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render a successful <Unregister />: success-unregister 1`] = `
<div
  data-testid="unregister-sw-success"
>
  <div
    class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomCenter css-1t20nyk-MuiSnackbar-root"
    color="primary"
    role="presentation"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation6 MuiSnackbarContent-root css-1eqdgzv-MuiPaper-root-MuiSnackbarContent-root"
      direction="up"
      role="alert"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiSnackbarContent-message css-1exqwzz-MuiSnackbarContent-message"
      >
        <div
          class="MuiBox-root css-1dhxv6h"
        >
          <div
            class="MuiBox-root css-oq1jtl"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ThumbUpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z"
              />
            </svg>
          </div>
          <div
            class="MuiBox-root css-1snmjsk"
          >
            Service worker loaded!
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render an error <Unregister />: error-unregister 1`] = `
<div
  data-testid="unregister-sw-err"
/>
`;
